cps/src/main/java/org/jenkinsci/plugins/workflow/cps/persistence/IteratorHack.java

```diff
@@ -0,0 +1,114 @@
+import java.util.AbstractList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import jenkins.model.Jenkins;
+import org.jenkinsci.plugins.workflow.pickles.Pickle;
+import org.jenkinsci.plugins.workflow.pickles.PickleFactory;
+
+/**
+ * Makes Java iterators effectively serializable.
+ */
+@Extension public class IteratorHack extends PickleFactory {
+
+    private static final Logger LOGGER = Logger.getLogger(IteratorHack.class.getName());
+    private static final Set<String> ACCEPTED_TYPES = ImmutableSet.of("java.util.ArrayList$Itr");
+
+    @Override public Pickle writeReplace(Object object) {
+        if (ACCEPTED_TYPES.contains(object.getClass().getName())) {
+            return new Replacement(object);
+        } else {
+            return null;
+        }
+    }
+
+    @SuppressWarnings("rawtypes")
+    private static class Replacement extends Pickle {
+
+        private final Class type;
+        private final Map<Class,Map<String,Object>> fields;
+
+        Replacement(Object o) {
+            type = o.getClass();
+            fields = new HashMap<>();
+            ReflectionProvider rp = Jenkins.XSTREAM2.getReflectionProvider();
+            rp.visitSerializableFields(o, new ReflectionProvider.Visitor() {
+                @Override public void visit(String name, Class type, Class definedIn, Object value) {
+                    if (name.equals("expectedModCount")) {
+                        LOGGER.log(Level.FINER, "ignoring expectedModCount={0}", value);
+                        return;
+                    }
+                    Map<String,Object> fieldsByClass = fields.get(definedIn);
+                    if (fieldsByClass == null) {
+                        fieldsByClass = new HashMap<>();
+                        fields.put(definedIn, fieldsByClass);
+                    }
+                    fieldsByClass.put(name, value);
+                }
+            });
+            LOGGER.log(Level.FINE, "replacing {0} with {1}", new Object[] {o, fields});
+        }
+
+        @Override public ListenableFuture<?> rehydrate() {
+            ReflectionProvider rp = Jenkins.XSTREAM2.getReflectionProvider();
+            Object o = rp.newInstance(type);
+            for (Map.Entry<Class,Map<String,Object>> entry : fields.entrySet()) {
+                Class definedIn = entry.getKey();
+                for (Map.Entry<String,Object> entry2 : entry.getValue().entrySet()) {
+                    String fieldName = entry2.getKey();
+                    Object value = entry2.getValue();
+                    rp.writeField(o, fieldName, value, definedIn);
+                    if (fieldName.equals("this$0")) {
+                        try {
+                            Field f = AbstractList.class.getDeclaredField("modCount"); // TODO if not handling only ArrayList.Itr, search in superclasses
+                            f.setAccessible(true);
```

Comment:
Any concern about security restrictions in different environments, this being an internal JVM class? Could you not just transform the list and/or iterator to an alternate serializable version?